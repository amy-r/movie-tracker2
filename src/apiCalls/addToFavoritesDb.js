export const addToFavoritesDb = async (movieData, userId) => {
  // eslint-disable-next-line
  const favoriteData = { ...movieData, user_id: userId };
  const url = "api/users/favorites/new";
  try {
    const response = await fetch(url, {
      method: "POST",
      body: JSON.stringify(favoriteData),
      headers: { "Content-Type": "application/json" }
    });
    const favoriteId = await response.json();
    return favoriteId;
  } catch (error) {
    throw new Error("Error adding to favorites");
  }
};
