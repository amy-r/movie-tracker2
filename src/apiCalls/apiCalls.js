import apiKey from './apiKey';
import { movieCleaner } from './movieCleaner'

export const url = `https://api.themoviedb.org/3/discover/movie?api_key=${apiKey}&primary_release_date.gte=2018-02-27&primary_release_date.lte=2018-03-27`;

export const getMovies = async (url) => {
  try {
    const response = await fetch(url);
    const movies = await response.json();
    const cleanMovies = movieCleaner(movies.results);
    return cleanMovies;
  } catch (error) {
    throw new Error('Error getting movies');
  }
};

export const userLogin = async credentials => {
  const url = 'api/users';
  try {
    const response = await fetch(url, {
      method: 'POST',
      body: JSON.stringify(credentials),
      headers: { 'Content-Type': 'application/json' }
    });
    const user = await response.json();
    return user.data;
  } catch (error) {
    throw error;
  }
};

export const userSignup = async accountInfo => {
  const url = 'api/users/new';
  try {
    const response = await fetch(url, {
      method: 'POST',
      body: JSON.stringify(accountInfo),
      headers: { 'Content-Type': 'application/json' }
    });
    const userId = await response.json();
    return userId;
  } catch (error) {
    throw error;
  }
};
