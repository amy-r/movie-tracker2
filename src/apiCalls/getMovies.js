const apiKey = process.env.REACT_APP_MDB_API_KEY

export const getMovies = async () => {
  const urlRoot = "https://api.themoviedb.org/3/movie/now_playing?";
  const urlKey = `api_key=${apiKey}`;
  const urlQuery = "&language=en-US&page=1";
  const url = `${urlRoot}${urlKey}${urlQuery}`;

  try {
    const response = await fetch(url);
    const movies = await response.json();
    return movies.results;
  } catch (error) {
    throw new Error("Error getting movies");
  }
};
