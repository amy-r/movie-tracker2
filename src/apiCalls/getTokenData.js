export const getTokenData = (token) => {
  const base64Url = token.split('.')[1];
  const base64 = base64Url.replace(/-/g, '+').replace(/_/g, '/');
  const jsonPayload = decodeURIComponent(
    window.atob(base64).split('').map(function (segment) {
      return '%' + ('00' + segment.charCodeAt(0).toString(16)).slice(-2);
    }).join('')
  );
  return JSON.parse(jsonPayload);
}