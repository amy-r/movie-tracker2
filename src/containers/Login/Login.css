.login {
  width: 500px;
  height: 250px;
  padding: 30px;
  margin: 160px auto 0 auto;
  background-color: #1481BA;
}

h4 {
  color: #f5f5f5;
  margin: 0px 20px 0px 20px;
}

form {
  margin: 30px auto;
  width: 320px;
}

input {
  display: block;
  padding-left: 5px;
  width: 300px;
  height: 35px;
  margin: 25px auto;
  font-size: 20px;
}

h2.errorMessage {
  color: #034748;
  margin-bottom: 0;
}

form button {
  font-size: 20px;
  width: 307px;
  height: 35px;
  background-color: #034748;
  color: #f5f5f5;
}

button:disabled {
  background-color: #11B5E4;
}

@media screen and (max-width: 720px){

  .login {
    width: 90%;
    height: 280px;
    padding: 10px;
  }

}

@media screen and (max-width: 500px){

  .login {
    padding: 10px;
  }
  
  input, form button {
    width: 200px;
    margin: 0 auto;
  }

  form {
  margin: auto;
  width: 90%;
}


}

