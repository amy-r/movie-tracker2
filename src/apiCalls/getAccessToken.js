import { getTokenData } from "./getTokenData";

export const getAccessToken = async () => {
  const accessToken = sessionStorage.getItem('accessToken');
  const decodedToken = accessToken ? getTokenData(accessToken) : null;
  const expirationTime = decodedToken ? decodedToken.exp : null;
  const now = Math.trunc(Date.now() / 1000);

  // Token is expired or doesn't exist
  if (expirationTime && now >= expirationTime) {
    const apiRoot = process.env.REACT_APP_AUTH_API_URL;
    const url = apiRoot + "/auth/refresh";
    try {
      const response = await fetch(url, {
        method: "GET",
        headers: { "Content-Type": "application/json" },
        credentials: "include"
      });
      const tokens = await response.json();
      sessionStorage.setItem("accessToken", tokens.accessToken);
      return tokens.accessToken;
    } catch (error) {
      throw new Error("Error getting access token");
    }
  // Token is not expired
  } else if (now < expirationTime) {
    return accessToken;
  }
};