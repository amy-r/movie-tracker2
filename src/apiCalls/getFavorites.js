/* eslint-disable camelcase */
import { getAccessToken } from "./getAccessToken";

export const getFavorites = async userId => {
  const token = await getAccessToken();
  if (!token) {
    return null;
  }
  const apiRoot = process.env.REACT_APP_AUTH_API_URL;
  const url = apiRoot + `/users/${userId}/favorites`;
  try {
    const response = await fetch(url, {
      method: "GET",
      headers: { 
        "Content-Type": "application/json",
        "Authorization": `Bearer ${token}`
      }
    });
    const favorites = await response.json();
    const cleanFavorites = favorites.data.favorites.map(favorite => {
      return {
        id: favorite.id,
        user_id: favorite.user_id,
        movie_id: parseInt(favorite.movie_id, 10),
        created_at: favorite.created_at
      };
    });
    return cleanFavorites;
  } catch (error) {
    throw new Error("Error getting favorites");
  }
};
