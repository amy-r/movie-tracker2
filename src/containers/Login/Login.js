import React, { Component } from 'react';
import { withRouter } from 'react-router-dom';
import { userLogin } from '../../apiCalls/userLogin';
import { getFavorites } from '../../apiCalls/getFavorites';
import { validateUser, setError, addFavorites } from '../../actions';
import './Login.css';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

export class Login extends Component {
  constructor() {
    super();
    this.state = {
      email: '',
      password: ''
    };
  }

  handleChange = event => {
    const { name, value } = event.target; 
    this.setState({ [name]: value });
  }

  handleSubmit = async event => {
    event.preventDefault();
    try {
      const user = await userLogin(this.state);
      const favorites = await getFavorites(user.id);
      this.props.validateUser(user);
      this.props.addFavorites(favorites);
      this.props.history.push('/');
    } catch (error) {
      this.setState({
        email: "",
        password: "",
        errorMessage: "Email and password do not match."
      });
    }
  }
  
  render() {
    const enableSubmit = this.state.email && this.state.password; 
    return (
      <div className="login">
        <h4>Please enter your email address and password to login.</h4>
        <form onSubmit={this.handleSubmit}>
          <input 
            type="text" 
            name="email" 
            value={this.state.email}
            placeholder="Email address" 
            onChange={this.handleChange} />
          <input 
            type="password"
            name="password"
            value={this.state.password}
            placeholder="Password"
            onChange={this.handleChange} />
          <button
            type="submit" 
            name="submit"
            disabled={!enableSubmit}>
              Submit
          </button>
        </form>
        <h2 className="errorMessage">{this.state.errorMessage}</h2>
      </div>
    );
  }
}

export const mapDispatchToProps = dispatch => ({
  validateUser: user => dispatch(validateUser(user)),
  setError: error => dispatch(setError(error)),
  addFavorites: favorites => dispatch(addFavorites(favorites))
});

Login.propTypes = {
  validateUser: PropTypes.func,
  setError: PropTypes.func,
  addFavorites: PropTypes.func,
  history: PropTypes.object
};

export default withRouter(connect(null, mapDispatchToProps)(Login));
