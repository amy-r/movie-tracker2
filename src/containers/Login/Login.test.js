import React from 'react';
import { Login, mapDispatchToProps } from './Login';
import { shallow } from 'enzyme';
import { userLogin } from '../../apiCalls/userLogin';
import { getFavorites } from '../../apiCalls/getFavorites';

jest.mock('../../apiCalls/userLogin');
jest.mock('../../apiCalls/getFavorites');

describe('Login', () => {

  let wrapper;
  const mockValidateUser = jest.fn();
  const mockAddFavorites = jest.fn();

  beforeEach(() => {
    wrapper = shallow(
      <Login
        name={''}
        email={''}
        validateUser={mockValidateUser}
        addFavorites={mockAddFavorites}
      />
    );
  });

  it('matches the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('loads with an emtpy state', () => {
    expect(wrapper.state('email')).toEqual('');
    expect(wrapper.state('password')).toEqual('');
  });


  it('updates the state as the user types', () => {
    const expected = {
      email: 'me@gmail.com',
      password: 'password'
    };
    const event1 = {
      target: {
        name: 'email',
        value: 'me@gmail.com'
      }
    };
    const event2 = {
      target: {
        name: 'password',
        value: 'password'
      }
    };
    wrapper.instance().handleChange(event1);
    wrapper.instance().handleChange(event2);
    expect(wrapper.state()).toEqual(expected);
  });

  it('calls userLogin with correct parameters on handleSubmit', () => {
    const expectedState = {
      email: 'me@gmail.com',
      password: 'password'
    };
    const event = { preventDefault: jest.fn() };
    wrapper.setState({
      email: 'me@gmail.com',
      password: 'password'
    });
    wrapper.instance().handleSubmit(event);
    expect(userLogin).toHaveBeenCalledWith(expectedState);
  });

  it('calls getFavorites on handleSubmit', () => {
    const event = { preventDefault: jest.fn() };
    wrapper.instance().handleSubmit(event);
    expect(getFavorites).toHaveBeenCalled();
  });

  it('calls validateUser on handleSubmit', () => {
    const event = { preventDefault: jest.fn() };
    wrapper.instance().handleSubmit(event);
    expect(mockValidateUser).toHaveBeenCalled();
  });

  it('calls addFavorites on handleSubmit', () => {
    const event = { preventDefault: jest.fn() };
    wrapper.instance().handleSubmit(event);
    expect(mockAddFavorites).toHaveBeenCalled();
  });

  it.skip('displays an error message when credentials dont match', () => {
    const expected = "Email and password do not match.";
    const event = { preventDefault: jest.fn() };
    window.fetch = jest.fn().mockImplementation(() =>
      Promise.reject({
        ok: false
      })
    );
    wrapper.instance().handleSubmit(event);
    expect(wrapper.state('errorMessage')).toEqual(expected);
  });

  it('disables submit button unless both fields contain data', () => {
    expect(wrapper.find('button').prop('disabled')).toBe(true);
    wrapper.setState({ email: 'me@gmail.com', password: ''});
    expect(wrapper.find('button').prop('disabled')).toBe(true);
    wrapper.setState({ email: 'me@gmail.com', password: 'password'});
    expect(wrapper.find('button').prop('disabled')).toBe(false);
  });

  it('should call dispatch function when using mapDispatchToProps', () => {
    const mockDispatch = jest.fn();
    const mapped = mapDispatchToProps(mockDispatch);
    mapped.setError();
    mapped.addFavorites();
    mapped.validateUser();
    expect(mockDispatch).toHaveBeenCalled();
  });


});