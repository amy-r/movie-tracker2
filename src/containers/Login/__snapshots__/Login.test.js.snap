// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login
    addFavorites={[Function]}
    email=""
    name=""
    validateUser={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h4>
          Please enter your email address and password to login.
</h4>,
        <form
          onSubmit={[Function]}
>
          <input
                    name="email"
                    onChange={[Function]}
                    placeholder="Email address"
                    type="text"
                    value=""
          />
          <input
                    name="password"
                    onChange={[Function]}
                    placeholder="Password"
                    type="password"
                    value=""
          />
          <button
                    disabled={true}
                    name="submit"
                    type="submit"
          >
                    Submit
          </button>
</form>,
        <h2
          className="errorMessage"
/>,
      ],
      "className": "login",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Please enter your email address and password to login.",
        },
        "ref": null,
        "rendered": "Please enter your email address and password to login.",
        "type": "h4",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              name="email"
              onChange={[Function]}
              placeholder="Email address"
              type="text"
              value=""
/>,
            <input
              name="password"
              onChange={[Function]}
              placeholder="Password"
              type="password"
              value=""
/>,
            <button
              disabled={true}
              name="submit"
              type="submit"
>
              Submit
</button>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "email",
              "onChange": [Function],
              "placeholder": "Email address",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "password",
              "onChange": [Function],
              "placeholder": "Password",
              "type": "password",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Submit",
              "disabled": true,
              "name": "submit",
              "type": "submit",
            },
            "ref": null,
            "rendered": "Submit",
            "type": "button",
          },
        ],
        "type": "form",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": undefined,
          "className": "errorMessage",
        },
        "ref": null,
        "rendered": null,
        "type": "h2",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h4>
            Please enter your email address and password to login.
</h4>,
          <form
            onSubmit={[Function]}
>
            <input
                        name="email"
                        onChange={[Function]}
                        placeholder="Email address"
                        type="text"
                        value=""
            />
            <input
                        name="password"
                        onChange={[Function]}
                        placeholder="Password"
                        type="password"
                        value=""
            />
            <button
                        disabled={true}
                        name="submit"
                        type="submit"
            >
                        Submit
            </button>
</form>,
          <h2
            className="errorMessage"
/>,
        ],
        "className": "login",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Please enter your email address and password to login.",
          },
          "ref": null,
          "rendered": "Please enter your email address and password to login.",
          "type": "h4",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                name="email"
                onChange={[Function]}
                placeholder="Email address"
                type="text"
                value=""
/>,
              <input
                name="password"
                onChange={[Function]}
                placeholder="Password"
                type="password"
                value=""
/>,
              <button
                disabled={true}
                name="submit"
                type="submit"
>
                Submit
</button>,
            ],
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "name": "email",
                "onChange": [Function],
                "placeholder": "Email address",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "name": "password",
                "onChange": [Function],
                "placeholder": "Password",
                "type": "password",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Submit",
                "disabled": true,
                "name": "submit",
                "type": "submit",
              },
              "ref": null,
              "rendered": "Submit",
              "type": "button",
            },
          ],
          "type": "form",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": undefined,
            "className": "errorMessage",
          },
          "ref": null,
          "rendered": null,
          "type": "h2",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
